[Bug Fix] Fix CUDA 13.x library loading and installation dependencies

youge325 opened the pull request:
<!-- TemplateReference: https://github.com/PaddlePaddle/Paddle/wiki/PULL-REQUEST-TEMPLATE--REFERENCE -->
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->

### PR Category
Environment Adaptation


### PR Types
Bug fixes


### Description
- dynamic_loader.cc：在 `PADDLE_WITH_PIP_CUDA_LIBRARIES` 场景下新增了 CUDA 13.x 分支，按需加载 `cusolver64_12.dll`、`cusparse64_12.dll`、`cufft64_12.dll` 等 Windows pip 包里提供的新文件名，避免 CUDA13 用户因旧分支只识别 11/12 版本而找不到库。
- setup.py.in：在 `get_paddle_extra_install_requirements()` 里补充了 `"13.0"` 对应的 NVIDIA pip 依赖清单（`nvidia-cuda-runtime-cu13`、`nvidia-cublas-cu13` 等），保证打包/安装 GPU whl 时会自动拉取 CUDA13 所需组件。
- 测试代码如下：
```python
import paddle
paddle.utils.run_check()
```
#### Before
```shell
C:\Users\Lenovo\Downloads\paddle>D:/Users/Lenovo/AppData/Local/Programs/Python/Python310/python.exe c:/Users/Lenovo/Downloads/paddle/demo13.py
Traceback (most recent call last):
  File "c:\Users\Lenovo\Downloads\paddle\demo13.py", line 1, in <module>
    import paddle
  File "D:\Users\Lenovo\AppData\Local\Programs\Python\Python310\lib\site-packages\paddle\__init__.py", line 919, in <module>
    raise err
OSError: [WinError 127] 找不到指定的程序。 Error loading "D:\Users\Lenovo\AppData\Local\Programs\Python\Python310\lib\site-packages\paddle\..\nvidia\cusolver\bin\cusolver64_11.dll" or one of its dependencies.
```
#### After
```shell
C:\Users\Lenovo\Downloads\paddle>D:/Users/Lenovo/AppData/Local/Programs/Python/Python310/python.exe c:/Users/Lenovo/Downloads/paddle/demo13.py
Running verify PaddlePaddle program ... 
D:\Users\Lenovo\AppData\Local\Programs\Python\Python310\lib\site-packages\paddle\pir\math_op_patch.py:240: UserWarning: Tensor do not have 'place' interface for pir graph mode, try not to use it. None will be returned.
  warnings.warn(
I1019 23:18:46.195433  8912 pir_interpreter.cc:1524] New Executor is Running ...
WARNING: Logging before InitGoogleLogging() is written to STDERR
W1019 23:18:46.195433  8912 gpu_resources.cc:114] Please NOTE: device: 0, GPU Compute Capability: 8.6, Driver API Version: 13.0, Runtime API Version: 13.0
I1019 23:18:46.202929  8912 pir_interpreter.cc:1547] pir interpreter is running by multi-thread mode ...
PaddlePaddle works well on 1 GPU.
PaddlePaddle is installed successfully! Let's start deep learning with PaddlePaddle now.
```
- issue: https://github.com/PaddlePaddle/Paddle/issues/75848

youge325 commented:
/re-run all-failed

youge325 commented:
/re-run all-failed

youge325 commented:
/re-run all-failed

youge325 committed:
[Bug Fix] Fix CUDA 13.x library loading and installation dependencies
- dynamic_loader.cc：在 `PADDLE_WITH_PIP_CUDA_LIBRARIES` 场景下新增了 CUDA 13.x 分支，按需加载 `cusolver64_12.dll`、`cusparse64_12.dll`、`cufft64_12.dll` 等 Windows pip 包里提供的新文件名，避免 CUDA13 用户因旧分支只识别 11/12 版本而找不到库。
- setup.py.in：在 `get_paddle_extra_install_requirements()` 里补充了 `"13.0"` 对应的 NVIDIA pip 依赖清单（`nvidia-cuda-runtime-cu13`、`nvidia-cublas-cu13` 等），保证打包/安装 GPU whl 时会自动拉取 CUDA13 所需组件。

youge325 committed:
Merge branch 'PaddlePaddle:develop' into dyn

swgu98 commented on paddle/phi/backends/dynload/dynamic_loader.cc:
```diff
@@ -721,8 +721,13 @@ void* GetCusolverDsoHandle() {
 #ifdef PADDLE_WITH_PIP_CUDA_LIBRARIES
-  return GetDsoHandleFromSearchPath(
-      FLAGS_cuda_dir, "cusolver64_11.dll", true, {cuda_lib_path});
+  if (CUDA_VERSION >= 13000 && CUDA_VERSION < 14000) {
```
cusolver大版本并不随cuda严格同步更新，建议和linux端逻辑保持一致，不要限制cuda13到cuda14

youge325 replied:
好的，已参考linux端逻辑进行修改

swgu98 commented on python/setup.py.in:
*-cu13包在pypi上并没有更新，这里还是严格限制下版本比较好，可以参考linux端版本测下

youge325 replied:
好的，已参考linux端版本进行修改，编译没有报错，安装时自动安装了对应版本的包，运行paddle.utils.run_check()没有报错

youge325 committed:
Merge branch 'PaddlePaddle:develop' into dyn

youge325 committed:
fix pip install for CUDA 13.0 on Windows and specify exact versions for CUDA 13.0 dependencies
\

**修改位置**：`GetCusolverDsoHandle()` 函数

**修改内容**：
- **修改前**：在 Windows + CUDA 环境下，逻辑判断有误 - 先检查 CUDA 13.x，再检查 < 13.x
- **修改后**：正确的逻辑顺序 - 先检查 CUDA < 13.0，再处理 >= 13.0 的情况

**具体修复**：
```cpp
// 修改前（错误的逻辑）
if (CUDA_VERSION >= 13000 && CUDA_VERSION < 14000) {
    return GetDsoHandleFromSearchPath(FLAGS_cuda_dir, "cusolver64_12.dll", ...);
} else {
    return GetDsoHandleFromSearchPath(FLAGS_cuda_dir, "cusolver64_11.dll", ...);
}

// 修改后（正确的逻辑）
if (CUDA_VERSION < 13000) {
    return GetDsoHandleFromSearchPath(FLAGS_cuda_dir, "cusolver64_11.dll", ...);
} else {
    return GetDsoHandleFromSearchPath(FLAGS_cuda_dir, "cusolver64_12.dll", ...);
}
```

**作用**：确保在 Windows 系统上正确加载对应 CUDA 版本的 `cusolver` 库文件。

**修改位置**：`get_paddle_extra_install_requirements()` 函数中的 CUDA 13.0 依赖项

**修改内容**：将 CUDA 13.x 的依赖包从"不指定版本"改为"指定具体版本号"

**具体修复**：
```python
"13.0": (
    "nvidia-cuda-runtime-cu13 | "
    "nvidia-cudnn-cu13 | "
    ...
)

"13.0": (
    "nvidia-cuda-runtime==13.0.88 | "
    "nvidia-cudnn-cu13==9.14.0.64 | "
    "nvidia-cublas==13.1.0.3 | "
    ...
)
```

**作用**：
- 确保通过 pip 安装时，能够正确安装 CUDA 13.x 所需的特定版本的 NVIDIA 库
- 避免安装不兼容的库版本
- 保证库之间的版本兼容性
